File: scripts/toolchain/fetch.py
#!/usr/bin/env python3
"""
Download and extract toolchain archives idempotently.

Usage:
  fetch.py --url <tarball-url> --archive <path/to/archive.tar.gz> \
           --dest <extract_dir> --expected <dir_created_by_extract>
"""

from __future__ import annotations

import argparse
import sys
import tarfile
import urllib.request
from pathlib import Path


def _parse_args(argv: list[str]) -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Fetch and extract toolchain archive")
    p.add_argument("--url", required=True, help="URL to download")
    p.add_argument("--archive", required=True, help="Local archive path")
    p.add_argument("--dest", required=True, help="Directory to extract into")
    p.add_argument("--expected", required=True, help="Directory that should exist after extraction")
    return p.parse_args(argv[1:])


def _download(url: str, archive: Path) -> None:
    archive.parent.mkdir(parents=True, exist_ok=True)
    if archive.exists():
        print(f"[fetch] archive already present: {archive}")
        return
    print(f"[fetch] downloading {url} -> {archive}")
    with urllib.request.urlopen(url) as resp, archive.open("wb") as out:
        chunk = resp.read(8192)
        while chunk:
            out.write(chunk)
            chunk = resp.read(8192)


def _extract(archive: Path, dest: Path, expected: Path) -> None:
    if expected.exists():
        print(f"[fetch] extract skipped (found {expected})")
        return
    print(f"[fetch] extracting {archive} -> {dest}")
    dest.mkdir(parents=True, exist_ok=True)
    with tarfile.open(archive, "r:*") as tar:
        tar.extractall(dest)


def main(argv: list[str]) -> int:
    args = _parse_args(argv)
    archive = Path(args.archive)
    dest = Path(args.dest)
    expected = Path(args.expected)

    _download(args.url, archive)
    _extract(archive, dest, expected)
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: scripts/toolchain/fetch.lua
#!/usr/bin/env lua
-- Download and extract toolchain archives idempotently.
--
-- Usage:
--   fetch.lua --url <tarball-url> --archive <path/to/archive.tar.gz> \
--            --dest <extract_dir> --expected <dir_created_by_extract>

local function sh_quote(s)
  return "'" .. tostring(s):gsub("'", [['"'"']]) .. "'"
end

local function exec_checked(cmd)
  io.stdout:write(cmd .. "\n")
  local ok, why, code = os.execute(cmd)
  if type(ok) == "number" then
    code = ok
    ok = (code == 0)
  elseif ok == true then
    code = 0
  elseif code == nil then
    code = 1
  end
  if not ok then
    error(("command failed (%s): %s"):format(tostring(code), cmd))
  end
end

local function path_exists(path)
  local cmd = "test -e " .. sh_quote(path)
  local ok, _, code = os.execute(cmd)
  if type(ok) == "number" then
    return ok == 0
  end
  return ok == true and (code == 0 or code == nil)
end

local function mkdir_p(dir)
  exec_checked("mkdir -p " .. sh_quote(dir))
end

local function dirname(path)
  local d = tostring(path):match("^(.*)/[^/]+$") or ""
  if d == "" then
    return "."
  end
  return d
end

local function parse_args(argv)
  local out = {}
  local i = 1
  while i <= #argv do
    local a = argv[i]
    if a == "--url" or a == "--archive" or a == "--dest" or a == "--expected" then
      local v = argv[i + 1]
      if not v then
        error("missing value for " .. a)
      end
      out[a:sub(3)] = v
      i = i + 2
    elseif a == "-h" or a == "--help" then
      out.help = true
      i = i + 1
    else
      error("unknown arg: " .. tostring(a))
    end
  end
  return out
end

local function usage()
  io.stderr:write([[
usage: fetch.lua --url <url> --archive <path> --dest <dir> --expected <path>
]])
end

local function main(argv)
  local args = parse_args(argv)
  if args.help then
    usage()
    return 0
  end
  if not args.url or not args.archive or not args.dest or not args.expected then
    usage()
    return 2
  end

  local archive = args.archive
  local dest = args.dest
  local expected = args.expected

  mkdir_p(dirname(archive))
  if path_exists(archive) then
    io.stdout:write("[fetch] archive already present: " .. archive .. "\n")
  else
    io.stdout:write(("[fetch] downloading %s -> %s\n"):format(args.url, archive))
    exec_checked("curl -L " .. sh_quote(args.url) .. " -o " .. sh_quote(archive))
  end

  if path_exists(expected) then
    io.stdout:write("[fetch] extract skipped (found " .. expected .. ")\n")
  else
    io.stdout:write(("[fetch] extracting %s -> %s\n"):format(archive, dest))
    mkdir_p(dest)
    exec_checked("tar -C " .. sh_quote(dest) .. " -xf " .. sh_quote(archive))
  end
  return 0
end

local ok, err = pcall(function()
  os.exit(main(arg))
end)
if not ok then
  io.stderr:write("[fetch] error: " .. tostring(err) .. "\n")
  os.exit(1)
end
